HarfangLab/harfanglab/iocs.py

```diff
@@ -0,0 +1,193 @@
+        """
+
+        seen_values: defaultdict[str, list[str]] = defaultdict(list)
+        results: dict[str, Any] = {"valid": [], "revoked": [], "expired": []}
+
+        for object in stix_objects:
+            # Extract value and type from pattern
+            self.log(message=f"object in stix_objects {str(object)}", level="debug")
+            indicators = stix_to_indicators(stix_object=object, supported_types_map=self.SUPPORTED_TYPES_MAP)
+            for indicator in indicators:
+                ioc_value = indicator["value"]
+                ioc_type = indicator["type"]
+                result = self.get_payload(value=ioc_value, type=ioc_type, args=args)
+
+                # Handle revoked objects
+                if object.get("revoked", False):
+                    results["revoked"].append(result)
+                    continue
+
+                # Ignore this IOC if the same value has already been processed
+                if ioc_value in seen_values[ioc_type]:
+                    continue
+
+                # Add a direct link in description if the data is originating from Sekoia.io
+                if "x_ic_observable_types" in object.keys():
+                    result["description"] = self.get_reference_url(object["id"])
+
+                # Compare expiration data if exists
+                valid_until = object.get("valid_until")
+                if valid_until:
+                    current_datetime = datetime.now(timezone.utc)
+                    valid_until_datetime = isoparse(valid_until)
+                    if valid_until_datetime < current_datetime:
+                        results["expired"].append(result)
+                        continue
+
+                seen_values[ioc_type].append(ioc_value)
+                results["valid"].append(result)
+
+        return results
+
+    def delete_indicator_by_ids(self, indicator_ids: list[str]) -> None:
+        self.log(f"Removing {len(indicator_ids)} indicators from HarfangLab", level="info")
+
+        for indicator_id in indicator_ids:
+            self.api_delete_indicator(indicator_id=indicator_id)
+
+    def remove_indicators(self, indicators: list) -> None:
+        ids_to_remove = []
+
+        if len(indicators) > 0:
+            for indicator in indicators:
+                response = self.api_search_indicator(
+                    indicator_value=indicator["value"], source_id=indicator["source_id"]
+                )
+                found_ids = [item["id"] for item in response.json().get("results", [])]
+                ids_to_remove.extend(found_ids)
+
+        if len(ids_to_remove) > 0:
+            self.delete_indicator_by_ids(indicator_ids=ids_to_remove)
+
+    def create_indicators(self, indicators: list) -> None:
+        if len(indicators) > 0:
+            self.log(f"Pushing {len(indicators)} new indicators to HarfangLab", level="info")
+            for ioc in indicators:
+                try:
+                    self.api_create_indicator(indicator=ioc)
```

Comment:
With `_handle_response_error`, if we face an 400 response, we will log an error then an info about the fact that the IoC exist. This will mislead the customer. I think we should move the call to `_handle_response_error` in the exception block below